sklearn/ensemble/gradient_boosting.py

```diff
@@ -1184,6 +1184,37 @@ def _validate_y(self, y):
         # Default implementation
         return y
 
+    def apply(self, X):
+        """Apply trees in the forest to X, return leaf indices.
+
+        Parameters
+        ----------
+        X : array-like or sparse matrix, shape = [n_samples, n_features]
+            The input samples. Internally, it will be converted to
+            ``dtype=np.float32`` and if a sparse matrix is provided
+            to a sparse ``csr_matrix``.
+
+        Returns
+        -------
+        X_leaves : array_like, shape = [n_samples, n_estimators]
+            For each datapoint x in X and for each tree in the forest,
+            return the index of the leaf x ends up in.
+        """
+
+        if self.estimators_ is None or len(self.estimators_) == 0:
```

Comment:
Maybe some code from https://github.com/jmschrei/scikit-learn/blob/gbt_apply/sklearn/ensemble/gradient_boosting.py#L1068 could be factored out in a `_validate_X_predict` method?